content_harvester/by_page.py

```diff
@@ -0,0 +1,397 @@
+            print(
+                f"[{collection_id}, {page_filename}, {calisphere_id}]: "
+                f"Media Source: {media_src}, Temp File: {media_src_file}, "
+                f"fsize: {os.path.getsize(media_src_file)}"
+            )
+        media_dest = self.harvest_media(
+            calisphere_id, media_src, media_src_file)
+        print(
+            f"[{collection_id}, {page_filename}, {calisphere_id}] "
+            f"Media Path: {media_dest}"
+        )
+
+        # Harvest Thumbnail File for Record
+        thumbnail_src = record.get('thumbnail_source')
+
+        # this makes it so we don't have to re-write isShownBy mappings for
+        # non-nuxeo sources
+        if isinstance(thumbnail_src) == str:
+            thumbnail_src = {'url': thumbnail_src}
+
+        thumbnail_src_file = self._download_src(thumbnail_src)
+        if thumbnail_src_file:
+            print(
+                f"[{collection_id}, {page_filename}, {calisphere_id}]: "
+                f"Thumb Source: {thumbnail_src}, Temp File: "
+                f"{thumbnail_src_file}, fsize: "
+                f"{os.path.getsize(thumbnail_src_file)}"
+            )
+        thumbnail_dest = self.harvest_thumbnail(
+            thumbnail_src, thumbnail_src_file)
+        print(
+            f"[{collection_id}, {page_filename}, {calisphere_id}] "
+            f"Thumbnail Path: {thumbnail_dest}"
+        )
+
+        if media_src_file:
+            os.remove(media_src_file)
+        if thumbnail_src_file:
+            os.remove(thumbnail_src_file)
+
+        # Recurse through the record's children (if any)
+        child_records = get_child_records(collection_id, calisphere_id, self.s3)
+
+        print(
+            f"[{collection_id}, {page_filename}, {calisphere_id}]: "
+            f"{len(child_records)} children found."
+        )
+
+        children_media = []
+        for child in child_records:
+            children_media.append(self.harvest(child))
+
+        record['thumbnail'] = thumbnail_dest
+        if media_dest:
+            record['media'] = media_dest
+        if children_media:
+            record['children'] = children_media
+        return record
+
+    def _download_src(self, content_src) -> str:
+        '''
+            download source file to local disk
+        '''
+        if not content_src:
+            return None
+
+        src_url = content_src.get('url')
+        # filename defaults to last part of the url (filter clause handles urls
+        # with a trailing slash where the [-1] last item is an empty string)
+        filename = content_src.get(
+            'filename',
+            list(filter(lambda x: bool(x), src_url.split('/')))[-1]
+        )
+
+        tmp_file_path = os.path.join('/tmp', filename)
+        if os.path.exists(tmp_file_path):
+            return tmp_file_path
+
+        # Weird how we have to use username/pass to hit this endpoint
+        # but we have to use auth token to hit API endpoint
```

Comment:
Yeah, hrm, I don't know if this is the case anymore since we upgraded Nuxeo and got rid of Shibboleth.